UefiTestingPkg/FunctionalSystemTests/MemoryProtectionTest/App/MemoryProtectionTestApp.c

```diff
@@ -996,51 +1038,88 @@ UefiPoolGuard (
         return UNIT_TEST_SKIPPED;
       }
 
-      PoolTest ((UINT64 *)ptr, AllocationSize);
+      Attributes = 0;
+      if (mDxeMps.HeapGuardPolicy.Fields.Direction == HEAP_GUARD_ALIGNED_TO_TAIL) {
+        PoolGuard = ALIGN_ADDRESS (((UINTN)ptr) + AllocationSize + (EFI_PAGE_SIZE - 1));
+      } else {
+        PoolGuard = ALIGN_ADDRESS (((UINTN)ptr) - (EFI_PAGE_SIZE - 1));
+      }
 
-      UT_ASSERT_FALSE (GetIgnoreNextEx ());
-      UT_ASSERT_NOT_EFI_ERROR (mNonstopModeProtocol->ResetPageAttributes ());
+      UT_ASSERT_NOT_EFI_ERROR (mMemoryAttributeProtocol->GetMemoryAttributes (mMemoryAttributeProtocol, PoolGuard, EFI_PAGE_SIZE, &Attributes));
+      UT_ASSERT_NOT_EQUAL (Attributes & EFI_MEMORY_RP, 0);
     }
+  } else if (MemoryProtectionContext.TestingMethod == MemoryProtectionTestClearFaults) {
+    UT_ASSERT_NOT_NULL (mNonstopModeProtocol);
+    for (Index = 0; Index < NUM_POOL_SIZES; Index++) {
+      UT_ASSERT_NOT_EFI_ERROR (ExPersistSetIgnoreNextPageFault ());
 
-    return UNIT_TEST_PASSED;
-  }
+      //
+      // Context.TestProgress indicates progress within this specific test.
+      // The test progressively allocates larger areas to test the guard on.
+      // These areas are defined in Pool.c as the 13 different sized chunks that are available
+      // for pool allocation.
+      //
+      // We need to indicate we are working on the next part of the test and save our progress.
+      //
+      AllocationSize = mPoolSizeTable[Index];
 
-  if (MemoryProtectionContext.TestProgress < NUM_POOL_SIZES) {
-    //
-    // Context.TestProgress indicates progress within this specific test.
-    // The test progressively allocates larger areas to test the guard on.
-    // These areas are defined in Pool.c as the 13 different sized chunks that are available
-    // for pool allocation.
-    //
-    // We need to indicate we are working on the next part of the test and save our progress.
-    //
-    AllocationSize = mPoolSizeTable[MemoryProtectionContext.TestProgress];
-    MemoryProtectionContext.TestProgress++;
-    SaveFrameworkState (&MemoryProtectionContext, sizeof (MEMORY_PROTECTION_TEST_CONTEXT));
-    SetBootNextDevice ();
+      //
+      // Memory type rHardwareNxProtections to the bitmask for the PcdHeapGuardPoolType,
+      // we need to RShift 1 to get it to reflect the correct EFI_MEMORY_TYPE.
+      //

```

Comment:
I'm not quite making the connection about this comment relevant to the code. Is it relevant to the code here?

Reply:
That comment was outdated - removed